rfc/0001-fiona-2-0-changes.md

```diff
@@ -0,0 +1,39 @@
+
+Instead of working on another backwards compatible minor version of Fiona (1.9), the Fiona team shall begin working on what would be a new major version with some breaking changes.
+
+The planned changes are:
+
+* Dropping support for Python versions < 3.5.
+* Dropping support for GDAL versions < 2.2.
+* Replacing the existing coordinate referencing dicts with a new CRS class (as in Rasterio).
+* Replacing the `"geometry"` in `feature["geometry"]` with a Cython extension class that could enhance serialization performance. Geometry objects would become immutable.
+
+## Details
+
+The Python project will no longer support Python versions < 3.5 after the end of 2019. Python 3.4 has been retired. We can stop testing with 3.4 and stop building wheels for 3.4. Python 2.7 will not be retired until Jan 1, 2020, but Numpy and many associated projects have already decided to drop support for Python 2.7 in new versions starting Jan 1, 2019: https://python3statement.org/. Fiona can follow suit. Not only do we cut the cost of testing and building wheels for 2.7, we can eliminate the compat module and backported dependencies such as argparse, ordereddict, enum34, and mock.
+
+Many Fiona users get their distributions from PyPI and Conda-forge, which provide very recent versions of GDAL, 2.3 or 2.4. Many others get theirs from Ubuntu-GIS, which provides GDAL 2.1 in the stable PPA and GDAL 2.2 in the unstable PPA for Xenial (16.04), the oldest supported LTS release. The Fiona project has no regular input from other packagers downstream. Cutting off support for GDAL versions < 2.2 lightens the project build matrix and allows the deletion of `fiona/_shim1.*`, `fiona/_shim2.*`, and many lines of code in setup.py.
+
+Copying Rasterio's CRS class to fiona.crs will allow Fiona to change to WKT as the canonical representation of spatial reference systems and gain accuracy and precision. CRS objects would be immutable, which is a breaking change. The CRS class would be a Cython extension class with a OGRSpatialRefenceH object at its core.
+
+Representing spatial features and geometries as GeoJSON-like dicts has served Fiona well, but many of the project's users are looking for a faster representation. Replacing, e.g., `{"type": "Polygon", "coordinates": [[[...]]]}` with a Cython extension object based on a OGRGeometryH object will permit serialization to WKB or GeoJSON as needed and also faster read/filter/write pipelines that could completely bypass the GeoJSON serialization that has been standard until now.
+
+## Considerations
+
+It would be appropriate to warn users about upcoming changes for CRS and geometry. Doing so is somewhat complicated by the fact that we're currently using standard dicts for each. We may be faced with replacing the dicts with a custom, mutable dict-like object so that we can raise a DeprecationWarning when its `__setitem__` method is called. Discussion of whether this warrants another minor release before 2.0 is warranted.
+
```

Comment:
Oh, here's some decent guidance from the semver project: https://semver.org/#spec-item-7 and https://semver.org/#how-should-i-handle-deprecating-functionality (also in https://github.com/semver/semver/commit/e0f985a7266ac2fdbc4076f159e1008926c6fe1c).

Having burned myself with breaking changes updates a few times by not pinning dependencies at a major version, I can't imagine that this won't happen to people with Fiona 2. A 1.9 release with warnings would be a kind thing to do.

Instead of mucking with the dicts as I suggested above, I propose to raise warnings at two kinds of places:

1. Wherever crs dicts are passed into functions or methods as arguments.
2. Wherever crs dicts are returned from functions or methods.

@snowman2 @snorfalorpagus yes?

Reply:
Sounds reasonable to me.

Reply:
> Wherever crs dicts are passed into functions or methods as arguments.

You mean eg in the `fiona.open` method when writing a dataset?

Reply:
@jorisvandenbossche yes.

Reply:
One concern is that, if you start raising such warnings before 2.0, a valid workflow such as opening a file, getting the `crs` object, and using this `crs` unmodified to write a file (the first example in the readme is also doing more or less that) will start to give (unnecessary) warnings. Given that the majority of users will probably not modify the crs dict in place, I am not sure it is worth raising a warning for all those users?

Reply:
@jorisvandenbossche good point. I've tried addressed this concern in my latest edit.